File: 39-combination-sum/39-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] can, int target) {
        Arrays.sort(can);
        List<List<Integer>>res=new ArrayList<>();
        solve(0,can,target,res,new ArrayList<>(),0);
        return res;
    }
    
    public void solve(int ind,int []can,int target,List<List<Integer>>res,List<Integer>temp,int sum){
        if(ind>=can.length){
            if(sum==target){
                res.add(new ArrayList<>(temp));
            }
            return;
        }
        if(can[ind]+sum<=target){
            sum+=can[ind];
            temp.add(can[ind]);
            solve(ind,can,target,res,temp,sum);
            temp.remove(temp.size()-1);
            sum-=can[ind];
        }
        solve(ind+1,can,target,res,temp,sum);
    }
}

File: 39-combination-sum/39-combination-sum.cpp
class Solution {
public:
    vector<vector<int>>res;
    
    void solve(int i,int n,vector<int> & nums,vector<int> & temp,int target){
        if(i==n){
            if(target==0)
            res.push_back(temp);
            return;
        }
        
            if(nums[i]<=target){
                temp.push_back(nums[i]);
                solve(i,n,nums,temp,target-nums[i]);
                temp.pop_back();
            }
            solve(i+1,n,nums,temp,target);
        
    }
    vector<vector<int>> combinationSum(vector<int>& c, int target) {
        int n=c.size();
        vector<int>temp;
        solve(0,n,c,temp,target);
        return res;
    }
};